File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.cpp
class Solution {
public:
    /* ALGORITHM
    1. First we will create a 2D vector grid of size m*n initialised to 0 
       (indicating ungruarded) and mark guard cells with 2 and wall cells with 3.
    2. Create a direction array, dir = {-1, 0, 1, 0, -1}
    3. Then, we will iterate over all the guards and for each guard we will do the following stuffs
        3.1. we will do a dfs call in all 4 directions and mark unguarded calls as guarded if we do not
             encounter a guard or a wall.
    4. Iterate over all the elements in grid and return the count of all the unguarded cells.
    */
    // Time Complexity: O(g*(m+n) + O(m*n))
    // space complexity: O(m*n)
    vector <int> dir = {-1, 0, 1, 0, -1};
    void dfs(vector<vector<int>> &grid, int direction, int x, int y) {
        if ((x < 0) or (x >= grid.size()) or (y < 0) or (y >= grid[0].size()) or (grid[x][y] == 2) or (grid[x][y] == 3)) {
            return;
        }
        grid[x][y] = 1;
        // Do DFS in UPWARD DIRECTION
        if (direction == 0) dfs(grid, direction, x-1, y);
        // Do DFS in RIGHT DIRECTION
        else if (direction == 1) dfs(grid, direction, x, y+1);
        // Do DFS in DOWNWARD DIRECTION
        else if (direction == 2) dfs(grid, direction, x+1, y);
        // Do DFS in LEFTWARD DIRECTION
        else dfs(grid, direction, x, y-1);
    }
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        vector<vector<int>> grid (m, vector<int> (n, 0));
        int count = 0;

        // Mark guard cells
        for (auto& guard : guards) {
            grid[guard[0]][guard[1]] = 2;
        }
        // Mark wall cells
        for (auto& wall : walls) {
            grid[wall[0]][wall[1]] = 3;
        }
        // Iterate over each guard
        for (auto& guard : guards) {
            for (int i = 0; i < 4; i++) {
                int x = guard[0] + dir[i];
                int y = guard[1] + dir[i+1];
                // Do DFS() in all 4 directions.
                dfs(grid, i, x, y);
            }
        }
        // Count all unguarded cells
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 0) count++;
            }
        }
        // Return the count
        return count;
    }
};

File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.py
class Solution:
    def __init__(self):
        # Direction arrays: dirx and diry for the four directions (Up, Right, Down, Left)
        self.dirx = [-1, 0, 1, 0]  # x-coordinate changes for Up, Right, Down, Left
        self.diry = [0, 1, 0, -1]  # y-coordinate changes for Up, Right, Down, Left

    def dfs(self, grid, direction, x, y):
        # Check if the current cell is out of bounds or contains a guard (2) or a wall (3)
        if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[0]) or grid[x][y] == 2 or grid[x][y] == 3:
            return
        
        # Mark the current cell as guarded
        grid[x][y] = 1

        # Perform DFS in the specified direction
        if direction == 0:
            self.dfs(grid, direction, x - 1, y)  # Up
        elif direction == 1:
            self.dfs(grid, direction, x, y + 1)  # Right
        elif direction == 2:
            self.dfs(grid, direction, x + 1, y)  # Down
        else:
            self.dfs(grid, direction, x, y - 1)  # Left

    def countUnguarded(self, m, n, guards, walls):
        # Create a grid of size m x n, initialized to 0 (unguarded)
        grid = [[0] * n for _ in range(m)]
        count = 0

        # Mark the guard cells with 2
        for guard in guards:
            grid[guard[0]][guard[1]] = 2
        
        # Mark the wall cells with 3
        for wall in walls:
            grid[wall[0]][wall[1]] = 3
        
        # Iterate over each guard and perform DFS in all 4 directions
        for guard in guards:
            for i in range(4):
                x = guard[0] + self.dirx[i]
                y = guard[1] + self.diry[i]
                # Perform DFS in the direction specified by i
                self.dfs(grid, i, x, y)
        
        # Count the number of unguarded cells (cells with value 0)
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    count += 1

        # Return the count of unguarded cells
        return count
